### PR TITLE
[Internal QA] [Workspace Feeds] Pull the workspace feed data from different place in domain_member NVP

### PR DESCRIPTION
**Pull request by VickyStash**
### Explanation of Change

This PR implements `getCompanyFeeds` function which returns both custom and direct feeds.

### Fixed Issues

$ https://github.com/Expensify/App/issues/52251
PROPOSAL: N/A

### Tests

- [x] Verify that no errors appear in the JS console

1. Open the app.
2. Go to the workspace with active custom feed and open Company cards page.
3. Make sure the list of the cards is displayed as expected.
4. Open any card details, make sure it looks as expected.
5. Try to assign one more card, make sure the flow works correctly.
6. In the console, past the next code:
```
Onyx.merge('sharedNVP_private_domain_member_YOUR_WORKSPACE_ACCOUNT_ID', {
    settings: {
        oAuthAccountDetails: {
            'oauth.chase.com': {
                accountList: ['CREDIT CARD...6607'],
                credentials: 'xxxxx',
                expiration: 1730998958,
            },
        },
    },
});
```
7. Open the feed switcher and switch to the Chase feed.
8. Repeat the **step 6**.
9. Start the assign card flow from the Chase bank, make sure you see the cards to assign.
10. Create a new workspace. Enable Company cards feature there and create a new custom feed. Make sure it's pending state looks as expected.

### Offline tests

Same, as in the `Tests` section.

### QA Steps

- [x] Verify that no errors appear in the JS console

1. Open the app.
2. Create a workspace and enable Company cards feature there.
3. Add Custom feed.
4. Assign a card inside the custom feed. Make sure it works as expected.
5. The list of cards should look as expected. Open any of the cards to make sure it looks ok.
6. Repeat steps 3-5 with a Direct feed.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/cd54ebef-b151-4743-85b3-9cd2d677c92e

https://github.com/user-attachments/assets/12964b0e-6cea-4ec3-9990-18fae11ceaa3

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/50a62732-ee26-4f4f-ba95-ea5ce55dbf86

https://github.com/user-attachments/assets/5d06631d-d75d-4686-ba20-e5b487ca476f

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/6bba84e7-f012-4b68-9e49-d22cfd36f822

https://github.com/user-attachments/assets/1891f32a-1435-460b-add4-5e378650e33c

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/6f1fae83-a344-45ce-a4ca-3b0bf24581e4

https://github.com/user-attachments/assets/b1689e4c-927d-467b-ba34-3a8a0ecfefa5

https://github.com/user-attachments/assets/dcf2a105-8391-46ed-96fe-a007f740ba47

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/0c66a482-2b24-4aea-82f1-63bf9dae5315

https://github.com/user-attachments/assets/0e48c8e7-b25a-4e6f-acc9-703817826231

</details>

### CODE REVIEW

**Comment by VickyStash** [comment]
@allgandalf I haven't recorded ios web (Safari) video cause I have some troubles with it. I regularly see the error (on the main branch too). I guess it's something with my setup after long weekends, but could you please check it as well.

https://github.com/user-attachments/assets/ebcd962b-f020-4080-8912-bef04bc37d34

**Review by allgandalf — CHANGES_REQUESTED** [review]
Okay, great job! few minor comments, but i am confident we can merge this one today!

File: src/libs/CardUtils.ts [hunk 1]
```diff
@@ -242,11 +242,22 @@ function getCardFeedIcon(cardFeed: CompanyCardFeed | typeof CONST.EXPENSIFY_CARD
...
+
+function removeExpensifyCardFromCompanyCards(cardFeeds: OnyxEntry<CardFeeds>): CompanyFeeds {
+    if (!cardFeeds) {
         return {};
     }
-    return Object.fromEntries(Object.entries(companyCards).filter(([key]) => key !== CONST.EXPENSIFY_CARD.BANK));
+
+    const companyCards = getCompanyFeeds(cardFeeds);
+
+    return Object.fromEntries(Object.entries(companyCards).filter(([key]) => key !== CONST.EXPENSIFY_CARD.BANK)) as CompanyFeeds;
```
[the remark this answers is not in the document]
Comment (VickyStash) [reply src/libs/CardUtils.ts]: It was one compatible with `CompanyFeeds,`, so I removed the assertion.

File: src/pages/workspace/members/WorkspaceMemberDetailsPage.tsx [hunk 2]
```diff
@@ -82,7 +82,7 @@ function WorkspaceMemberDetailsPage({personalDetails, policy, route}: WorkspaceM
     const isCurrentUserOwner = policy?.owner === currentUserPersonalDetails?.login;
     const ownerDetails = personalDetails?.[policy?.ownerAccountID ?? -1] ?? ({} as PersonalDetails);
     const policyOwnerDisplayName = ownerDetails.displayName ?? policy?.owner ?? '';
-    const hasMultipleFeeds = Object.values(cardFeeds?.settings?.companyCards ?? {}).filter((feed) => !feed.pending).length > 0;
+    const hasMultipleFeeds = Object.values(CardUtils.getCompanyFeeds(cardFeeds)).filter((feed) => !feed.pending).length > 0;
```
[the remark this answers is not in the document]
Comment (VickyStash) [reply src/pages/workspace/members/WorkspaceMemberDetailsPage.tsx:85]: Later, this variable is used to check whether the company card should be shown for the member or can be assigned to the member.
There are no cards in the pending feed, and no card can be assigned. So this check makes sense on this screen.

File: src/pages/workspace/companyCards/WorkspaceCompanyCardsSettingsPage.tsx [hunk 3]
```diff
@@ -102,7 +110,7 @@ function WorkspaceCompanyCardsSettingsPage({
                         isVisible={deleteCompanyCardConfirmModalVisible}
                         onConfirm={deleteCompanyCardFeed}
                         onCancel={() => setDeleteCompanyCardConfirmModalVisible(false)}
-                        title={translate('workspace.moreFeatures.companyCards.removeCardFeedTitle', {feedName})}
+                        title={feedName && translate('workspace.moreFeatures.companyCards.removeCardFeedTitle', {feedName})}
```
[the remark this answers is not in the document]
Comment (VickyStash) [reply src/pages/workspace/companyCards/WorkspaceCompanyCardsSettingsPage.tsx:112]: Overall it shouldn't be the case. But even if it's so, the title is supported to be undefined.

File: src/libs/CardUtils.ts [hunk 4]
```diff
@@ -242,11 +242,22 @@ function getCardFeedIcon(cardFeed: CompanyCardFeed | typeof CONST.EXPENSIFY_CARD
     return Illustrations.AmexCompanyCards;
 }
 
-function removeExpensifyCardFromCompanyCards(companyCards?: Record<string, CardFeedData>) {
-    if (!companyCards) {
+function isCustomFeed(feed: CompanyCardFeed): boolean {
+    return CONST.COMPANY_CARD.FEED_BANK_NAME.MASTER_CARD === feed || CONST.COMPANY_CARD.FEED_BANK_NAME.VISA === feed || CONST.COMPANY_CARD.FEED_BANK_NAME.AMEX === feed;
```
[the remark this answers is not in the document]
Comment (VickyStash) [reply src/libs/CardUtils.ts:246]: Using `includes` results in TS error.
As an alternative we can use `some`, what do you think?
```suggestion
    return [CONST.COMPANY_CARD.FEED_BANK_NAME.MASTER_CARD, CONST.COMPANY_CARD.FEED_BANK_NAME.VISA, CONST.COMPANY_CARD.FEED_BANK_NAME.AMEX].some((value) => value === feed);

```

**Comment by VickyStash** [comment]
@allgandalf I think I don't have rights to update labels 🤔

**Comment by mountiny** [comment]
Making a build

File: src/libs/CardUtils.ts [hunk 5]
```diff
@@ -282,6 +293,17 @@ const getBankCardDetailsImage = (bank: ValueOf<typeof CONST.COMPANY_CARDS.BANKS>
...
 };
 
+function getCustomOrFormattedFeedName(feed?: CompanyCardFeed, companyCardNicknames?: CompanyCardNicknames): string | undefined {
+    if (!feed) {
+        return;
+    }
+
+    const customFeedName = companyCardNicknames?.[feed];
+    const formattedFeedName = Localize.translateLocal('workspace.companyCards.feedName', {feedName: getCardFeedName(feed)});
+
```
Comment (mountiny) [inline src/libs/CardUtils.ts:303]:
```suggestion
```

File: src/libs/CardUtils.ts [hunk 6]
```diff
@@ -242,10 +242,21 @@ function getCardFeedIcon(cardFeed: CompanyCardFeed | typeof CONST.EXPENSIFY_CARD
...
+    return {...cardFeeds?.settings?.companyCards, ...cardFeeds?.settings?.oAuthAccountDetails};
+}
+
+function removeExpensifyCardFromCompanyCards(cardFeeds: OnyxEntry<CardFeeds>): CompanyFeeds {
+    if (!cardFeeds) {
         return {};
     }
+
+    const companyCards = getCompanyFeeds(cardFeeds);
+
```
Comment (mountiny) [inline src/libs/CardUtils.ts:259]:
```suggestion
```

File: src/types/onyx/CardFeeds.ts
Comment (mountiny) [inline src/types/onyx/CardFeeds.ts:1]: Nice!

File: tests/unit/CardUtilsTest.ts [hunk 7]
```diff
@@ -9,44 +13,194 @@ const longDateHyphen = '09-2024';
...
+            expect(selectedFeed).toBe(CONST.COMPANY_CARD.FEED_BANK_NAME.MASTER_CARD);
+        });
+
+        it('Should return undefined if lastSelectedFeed is undefined and there is no card feeds', () => {
+            const lastSelectedFeed = undefined;
+            const cardFeeds = undefined;
+            const selectedFeed = CardUtils.getSelectedFeed(lastSelectedFeed, cardFeeds);
+            expect(selectedFeed).toBe(undefined);
+        });
     });
```
Comment (mountiny) [inline tests/unit/CardUtilsTest.ts:182]: Could you also test selecting the direct feed please?

File: src/pages/workspace/companyCards/WorkspaceCompanyCardsPage.tsx [hunk 8]
```diff
@@ -35,10 +35,10 @@ function WorkspaceCompanyCardPage({route}: WorkspaceCompanyCardPageProps) {
     const selectedFeed = CardUtils.getSelectedFeed(lastSelectedFeed, cardFeeds);
     const [cardsList] = useOnyx(`${ONYXKEYS.COLLECTION.WORKSPACE_CARDS_LIST}${workspaceAccountID}_${selectedFeed}`);
 
-    const companyCards = CardUtils.removeExpensifyCardFromCompanyCards(cardFeeds?.settings?.companyCards);
+    const companyCards = CardUtils.removeExpensifyCardFromCompanyCards(cardFeeds);
     const isLoading = !cardFeeds || !!(cardFeeds.isLoading && !companyCards);
-    const selectedCompanyCard = companyCards[selectedFeed ?? ''] ?? cardFeeds?.settings?.oAuthAccountDetails?.[selectedFeed ?? ''] ?? null;
-    const isNoFeed = isEmptyObject(companyCards) && isEmptyObject(cardFeeds?.settings?.oAuthAccountDetails) && !selectedCompanyCard;
+    const selectedCompanyCard = selectedFeed && companyCards[selectedFeed];
```
Comment (mountiny) [inline src/pages/workspace/companyCards/WorkspaceCompanyCardsPage.tsx:40]: I am confused by this property name, its the companyCards settings object, right? we should name it better
```suggestion
    const selectedCompanyCardSettings = selectedFeed && companyCards[selectedFeed];
```

File: src/libs/actions/CompanyCards.ts [hunk 9]
```diff
@@ -121,19 +122,20 @@ function setWorkspaceCompanyCardFeedName(policyID: string, workspaceAccountID: n
     API.write(WRITE_COMMANDS.SET_COMPANY_CARD_FEED_NAME, parameters, onyxData);
 }
 
-function setWorkspaceCompanyCardTransactionLiability(workspaceAccountID: number, policyID: string, bankName: string, liabilityType: string) {
+function setWorkspaceCompanyCardTransactionLiability(workspaceAccountID: number, policyID: string, bankName: CompanyCardFeed, liabilityType: string) {
     const authToken = NetworkStore.getAuthToken();
+    const isCustomFeed = CardUtils.isCustomFeed(bankName);
+    const feedUpdates = {
+        [bankName]: {liabilityType},
+    };
```
Comment (mountiny) [inline src/libs/actions/CompanyCards.ts:130]: seems like you handled the updates three different ways on this PR :D could you please use just one, for example the one you have in here setWorkspaceCompanyCardTransactionLiability. ie you define `feedUpdates` and then

`settings: isCustomFeed ? {companyCards: feedUpdates} : {oAuthAccountDetails: feedUpdates},`

**Review by mountiny — COMMENTED** [review]
Couple small changes requested

File: src/pages/workspace/companyCards/WorkspaceCompanyCardsPage.tsx (hunk 8 above)
Comment (VickyStash) [reply src/pages/workspace/companyCards/WorkspaceCompanyCardsPage.tsx]: It's not really a setting, more like a data object. So I've renamed it to `selectedFeedData`, let me know if you feel differently.

File: tests/unit/CardUtilsTest.ts (hunk 7 above)
Comment (VickyStash) [reply tests/unit/CardUtilsTest.ts]: Done https://github.com/Expensify/App/pull/52268/commits/07cb200e2800182477f89039910a56bdd486f0a6

File: src/libs/actions/CompanyCards.ts (hunk 9 above)
Comment (VickyStash) [reply src/libs/actions/CompanyCards.ts:130]: Updated!

**Review by allgandalf — APPROVED** [review]
LGTM and tests well, great job with the cleanup @VickyStash 🙇

**Comment by VickyStash** [comment]
@mountiny All yours! I've applied your feedback

**Review by mountiny — APPROVED** [review]
Thanks! great job!

**Comment by joekaufmanexpensify** [comment]
@VickyStash I am trying to QA this one, am I supposed to add a chase feed, or is the console snippet you added supposed to show one to me? I tried doing step 6, but don't see a Chase feed.

**Comment by VickyStash** [comment]
@joekaufmanexpensify I think the `Tests` section was more for the C+ testing, I think you should have the opportunity to test it all over real-life flow, so follow the `QA Steps` section. It can be any Direct feed, not Chase specifically.

**Pull request merged**